internal/parse/parser.go

```diff
@@ -187,57 +209,143 @@ func isNameByte(c byte) bool {
+// construct, if they n't parse they return false, restore the parser and leave
+// the default value in  other return type
+
+func (p *Parser) parseIdentifier() (string, bool) {
+	p.skipSpaces()
+	if p.pos >= len(p.input) {
+		return "", false
+	}
+	if p.peekByte('*') {
+		p.pos++
+		return "*", true
+	}
+
+	idStart := p.pos
+	if !isNameByte(p.input[p.pos]) {
+		return "", false
+	}
+	var i int
+	for i = p.pos; i < len(p.input); i++ {
+		if !isNameByte(p.input[i]) {
+			break
+		}
+	}
+	p.pos = i
+	return p.input[idStart:i], true
+}
+
+// Parses a column name or a Go type name. If parsing a Go type name then
+// struct name is in FullName.Prefix and the field name (if extant) is in
+// FullName.Name.
+// When parsing a column the table name (if extant) is in FullName.Prefix and
+// the column name is in FullName.Name func (p *Parser)
+func (p *Parser) parseFullName(idc idClass) (FullName, bool) {
+	cp := p.save()
+	var fn FullName
+	p.skipSpaces()
+	if id, ok := p.parseIdentifier(); ok {
+		fn.Prefix = id
+		if p.skipByte('.') {
+			if id, ok := p.parseIdentifier(); ok {
+				fn.Name = id
+				return fn, true
+			}
+		} else {
+			// A column name specified without a table prefix is a name not a
+			// prefix
+			if idc == columnId {
+				fn.Name = fn.Prefix
+				fn.Prefix = ""
+			}
+			return fn, true
+		}
+	}
+	cp.restore()
+	return fn, false
 }
 
-func TestRunTable(t *testing.T) {
-	var p = NewParser()
-	var parseTests = []parseHelperTest{
+// parseColumns parses text in the SQL query of the form "table.colname". If
+// there is more than one column then the columns must be bracketed together,
+// e.g.  "(col1, col2) AS Person".
+func (p *Parser) parseColumns() ([]FullName, bool) {
+	var cols []FullName
 
-		{bytef: p.peekByte, result: []bool{false}, input: "", data: []string{"a"}},
-		{bytef: p.peekByte, result: []bool{false}, input: "b", data: []string{"a"}},
-		{bytef: p.peekByte, result: []bool{true}, input: "a", data: []string{"a"}},
+	p.skipSpaces()
 
-		{bytef: p.skipByte, result: []bool{false}, input: "", data: []string{"a"}},
-		{bytef: p.skipByte, result: []bool{false}, input: "abc", data: []string{"b"}},
-		{bytef: p.skipByte, result: []bool{true, true}, input: "abc", data: []string{"a", "b"}},
+	// Case 1: A single column.
+	if col, ok := p.parseFullName(columnId); ok {
+		cols = append(cols, col)
 
-		{bytef: p.skipByteFind, result: []bool{false}, input: "", data: []string{"a"}},
-		{bytef: p.skipByteFind, result: []bool{false, true, true}, input: "abcde", data: []string{"x", "b", "c"}},
-		{bytef: p.skipByteFind, result: []bool{true, false}, input: "abcde ", data: []string{" ", " "}},
+		// Case 2: Multiple columns.
+	} else if p.skipByte('(') {
+		col, ok := p.parseFullName(columnId)
+		cols = append(cols, col)
+		p.skipSpaces()
+		// If the column names are not formated in a recognisable way then give
+		// up trying to parse.
+		if !ok {
+			return cols, false
```

Comment:
Same comment about saving the state from earlier above.

Reply:
parseColumns removed from this PR